osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/firmware/firmwarefile/FirmwareFile.java

```diff
@@ -0,0 +1,129 @@
+import lombok.extern.slf4j.Slf4j;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.ProtocolAdapterException;
+
+/**
+ * This Firmware file is defined according to the Smart Meter Requirements 5.1, Supplement 4, P2
+ * Companion Standard, Section 7. Firmware Upgrade
+ *
+ * <p>Fixed values are
+ *
+ * <ul>
+ *   <li>Firmware image magic number: 53h 4Dh 52h 35h
+ *   <li>Header Version: 00h
+ */
+@Slf4j
+public class FirmwareFile {
+
+  private byte[] imageData;
+
+  public static final String FIRMWARE_IMAGE_MAGIC_NUMBER = "534d5235";
+  public static final int HEADER_VERSION = 0;
+
+  public FirmwareFile(final byte[] imageData) {
+    this.imageData = imageData;
+    final Integer firmwareImageLength = this.getHeader().getFirmwareImageLengthInt();
+    final Integer securityLength = this.getHeader().getSecurityLengthInt();
+    final Integer headerLength = this.getHeader().getHeaderLengthInt();
+    if (imageData.length != (firmwareImageLength + securityLength + headerLength)) {
+      log.warn(
+          "Byte array length doesn't match lengths defined in header: "
+              + "\nByte array length : {}"
+              + "\nLengths defined in header: "
+              + "\nHeader : {}"
+              + "\nFirmwareImage : {}"
+              + "\nSecurity : {}"
+              + "\nTotal of {}  bytes.",
+          imageData.length,
+          headerLength,
+          firmwareImageLength,
+          securityLength,
+          (firmwareImageLength + securityLength + headerLength));
+    }
+  }
+
+  public int length() {
+    return this.imageData.length;
+  }
+
+  public byte[] getByteArray() {
+    return this.imageData;
+  }
+
+  public byte[] getHeaderByteArray() {
+    return this.readBytes(this.imageData, 0, this.getHeader().getHeaderLengthInt() - 1);
+  }
+
+  public byte[] getFirmwareImageByteArray() {
+    return this.readBytes(
+        this.imageData,
+        this.getHeader().getHeaderLengthInt(),
+        this.getHeader().getHeaderLengthInt() + this.getHeader().getFirmwareImageLengthInt() - 1);
+  }
+
+  public void setSecurityByteArray(final byte[] security) throws ProtocolAdapterException {
+    if (security.length != this.getHeader().getSecurityLengthInt()) {
+      throw new ProtocolAdapterException(
+          "Length of generated security byte array differs from length defined in header of firmwarefile");
+    }
+    this.imageData =
+        ByteBuffer.allocate(this.length())
+            .put(this.getHeaderByteArray())
+            .put(this.getFirmwareImageByteArray())
+            .put(security)
+            .array();
+  }
+
+  public void setMbusDeviceSerialNumber(final int mbusDeviceSerialNumber) {
+
+    final byte[] mbusDeviceSerialNumberByteArray =
+        ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN).putInt(mbusDeviceSerialNumber).array();
+    final ByteBuffer buffer = ByteBuffer.wrap(this.imageData);
+    buffer.position(22);
+    buffer.put(mbusDeviceSerialNumberByteArray);
```

Comment:
This puts the M-Bus device serial bytes into the firmware file without looking at the wildcard.
As I understand the documentation the wildcard does not have to be for the entire number, and part of it can be fixed (firmware for meters with IDs `00000001` to `00009999` with pattern `FFFF0000`).
Should it not be some kind of validation error if you would put a serial number that had values other than `0` in the positions where the pattern has `0` (more general: other than the specified char where the pattern does not have wildcard `F`)?

Reply:
Added check on wildcard on method setMbusDeviceSerialNumber()